File: rdps.c
/* rdps.c
 *
 * $Id$
 * 
 * Wireshark - Network traffic analyzer
 * By Gerald Combs <gerald@wireshark.org>
 * Copyright 1998 Gerald Combs
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

/* takes the file listed as the first argument and creates the file listed
as the second argument. It takes a PostScript file and creates a C program
with 2 functions:
	print_ps_preamble()
	print_ps_finale()

*/

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFFER_SIZE 1024

void start_code(FILE *fd, const char *func);
void write_code(FILE *fd, char *string);
void end_code(FILE *fd);
void ps_clean_string(char *out, const char *in,
			int outbuf_size);

enum ps_state { null, preamble, hex, finale };

int main(int argc, char **argv)
{
	FILE	*input;
	FILE	*output;
	char	buf[BUFFER_SIZE];	/* static sized buffer! */
	enum ps_state	state = null;

	if (argc != 3) {
		fprintf(stderr, "%s: input_file output_file\n", argv[0]);
		exit(-1);
	}

	if (!(input = fopen(argv[1], "r"))) {
		fprintf(stderr, "%s: cannot open %s for input.\n", argv[0], argv[1]);
		exit(-1);
	}

	if (!(output = fopen(argv[2], "w"))) {
		fprintf(stderr, "%s: cannot open %s for output.\n", argv[0], argv[2]);
		exit(-1);
	}

	fprintf(output, "/* Created by rdps.c. Do not edit! */\n\n"
          "#include <stdio.h>\n\n"
          "#include \"ps.h\"\n\n");

	while (fgets(buf, BUFFER_SIZE - 1, input)) {

		if (state == null) {
			if (strcmp(buf, "% ---- ethereal preamble start ---- %\n") == 0) {
				state = preamble;
				start_code(output, "preamble");
				continue;
			}
			else if (strcmp(buf, "% ---- ethereal finale start ---- %\n") == 0) {
				state = finale;
				start_code(output, "finale");
				continue;
			}
		}
		else if (state == preamble) {
			if (strcmp(buf, "% ---- ethereal preamble end ---- %\n") == 0) {
				state = null;
				end_code(output);
				continue;
			}
			else {
				write_code(output, buf);
			}
		}
		else if (state == hex) {
			if (strcmp(buf, "% ---- ethereal hex end ---- %\n") == 0) {
				state = null;
				end_code(output);
				continue;
			}
			else {
				write_code(output, buf);
			}
		}
		else if (state == finale) {
			if (strcmp(buf, "% ---- ethereal finale end ---- %\n") == 0) {
				state = null;
				end_code(output);
				continue;
			}
			else {
				write_code(output, buf);
			}
		}
		else {
			fprintf(stderr, "NO MATCH:%s", buf);
			exit(-1);
		}
	}
        exit(0);
}

void start_code(FILE *fd, const char *func)
{
	fprintf(fd, "/* Created by rdps.c. Do not edit! */\n");
	fprintf(fd, "void print_ps_%s(FILE *fd) {\n", func);
}

void write_code(FILE *fd, char *string)
{
	char psbuf[BUFFER_SIZE];
	ps_clean_string(psbuf, string, BUFFER_SIZE);
	fprintf(fd, "\tfprintf(fd, \"%s\");\n", psbuf);
}

void end_code(FILE *fd)
{
	fprintf(fd, "}\n\n\n");
}

void ps_clean_string(char *out, const char *in,
			int outbuf_size)
{
	int rd, wr;
	char c;

	for (rd = 0, wr = 0 ; wr < outbuf_size; rd++, wr++ ) {
		c = in[rd];
		switch (c) {
			case '\\':
				out[wr] = '\\';
				out[++wr] = '\\';
				out[++wr] = c;
				break;

			case '%':
				out[wr] = '%';
				out[++wr] = '%';
				break;

			case '\n':
				out[wr] = '\\';
				out[++wr] = 'n';
				break;

			default:
				out[wr] = c;
				break;
		}

		if (c == 0) {
			break;
		}
	}
}

File: tools/rdps.py
#!/usr/bin/env python3
#
# rdps.py
#
# Wireshark - Network traffic analyzer
# By Gerald Combs <gerald@wireshark.org>
# Copyright 1998 Gerald Combs
#
# SPDX-License-Identifier: GPL-2.0-or-later
#

'''\
takes the file listed as the first argument and creates the file listed
as the second argument. It takes a PostScript file and creates a C source
with 2 functions:
	print_ps_preamble()
	print_ps_finale()

Ported to Python from rdps.c.
'''

import sys
import os.path


def ps_clean_string(raw_str):
    ps_str = ''
    for c in raw_str:
        if c == '\\':
            ps_str += '\\\\'
        elif c == '\n':
            ps_str += '\\n'
        else:
            ps_str += c
    return ps_str


def start_code(fd, name):
    fd.write("static const char ps_%s[] =\n" % name)
    

def write_code(fd, raw_str):
    ps_str = ps_clean_string(raw_str)
    fd.write("\t\"%s\"\n" % ps_str)


def end_code(fd, name):
    fd.write(";\n")
    fd.write("\n")
    fd.write("void print_ps_%s(FILE *fd) {\n" % name)
    fd.write("\tfwrite(ps_%s, sizeof ps_%s - 1, 1, fd);\n" % ( name, name ) )
    fd.write("}\n\n\n")


def exit_err(msg=None, *param):
    if msg is not None:
        sys.stderr.write(msg % param)
    sys.exit(1)


# Globals
STATE_NULL = 'null'
STATE_PREAMBLE = 'preamble'
STATE_FINALE = 'finale'


def main():
    state = STATE_NULL

    if len(sys.argv) != 3:
        exit_err("%s: input_file output_file\n", __file__)

    input = open(sys.argv[1], 'r')
    output = open(sys.argv[2], 'w')

    script_name = os.path.split(__file__)[-1]

    output.write('''\
/* DO NOT EDIT
 *
 * Created by %s.
 *
 * ps.c
 * Definitions for generating PostScript(R) packet output.
 *
 * Wireshark - Network traffic analyzer
 * By Gerald Combs <gerald@wireshark.org>
 * Copyright 1998 Gerald Combs
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 */

#include <stdio.h>

#include "ps.h"

''' % script_name)

    for line in input:
        #line = line.rstrip()
        if state == STATE_NULL:
            if line.startswith("% ---- wireshark preamble start ---- %"):
                state = STATE_PREAMBLE
                start_code(output, "preamble")
                continue
            elif line.startswith("% ---- wireshark finale start ---- %"):
                state = STATE_FINALE
                start_code(output, "finale")
                continue
        elif state == STATE_PREAMBLE:
            if line.startswith("% ---- wireshark preamble end ---- %"):
                state = STATE_NULL
                end_code(output, "preamble")
                continue
            else:
                write_code(output, line)
        elif state == STATE_FINALE:
            if line.startswith("% ---- wireshark finale end ---- %"):
                state = STATE_NULL
                end_code(output, "finale")
                continue
            else:
                write_code(output, line)
        else:
            exit_err("NO MATCH:%s", line)

    sys.exit(0)

if __name__ == "__main__":
    main()

#
# Editor modelines  -  https://www.wireshark.org/tools/modelines.html
#
# Local variables:
# c-basic-offset: 4
# indent-tabs-mode: nil
# End:
#
# vi: set shiftwidth=4 expandtab:
# :indentSize=4:noTabs=true:
#
